graph/graph.go

```diff
@@ -428,10 +427,125 @@ func (graph *Graph) Heads() (map[string]*image.Image, error) {
 	return heads, err
 }
 
-func (graph *Graph) ImageRoot(id string) string {
+func (graph *Graph) imageRoot(id string) string {
 	return filepath.Join(graph.Root, id)
 }
 
-func (graph *Graph) Driver() graphdriver.Driver {
-	return graph.driver
+// storeImage stores file system layer data for the given image to the
+// graph's storage driver. Image metadata is stored in a file
+// at the specified root directory.
+func (graph *Graph) storeImage(img *image.Image, layerData archive.ArchiveReader, root string) (err error) {
+	// Store the layer. If layerData is not nil, unpack it into the new layer
+	if layerData != nil {
+		if img.Size, err = graph.driver.ApplyDiff(img.ID, img.Parent, layerData); err != nil {
+			return err
+		}
+	}
+
+	if err := graph.saveSize(root, int(img.Size)); err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(jsonPath(root), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(0600))
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	return json.NewEncoder(f).Encode(img)
+}
+
+// loadImage fetches the image with the given id from the graph.
+func (graph *Graph) loadImage(id string) (*image.Image, error) {
+	root := graph.imageRoot(id)
+
+	// Open the JSON file to decode by streaming
+	jsonSource, err := os.Open(jsonPath(root))
+	if err != nil {
+		return nil, err
+	}
+	defer jsonSource.Close()
+
+	img := &image.Image{}
+	dec := json.NewDecoder(jsonSource)
+
+	// Decode the JSON data
+	if err := dec.Decode(img); err != nil {
+		return nil, err
+	}
+	if err := image.ValidateID(img.ID); err != nil {
+		return nil, err
+	}
+
+	if buf, err := ioutil.ReadFile(filepath.Join(root, "layersize")); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		// If the layersize file does not exist then set the size to a negative number
+		// because a layer size of 0 (zero) is valid
+		img.Size = -1
+	} else {
+		// Using Atoi here instead would temporarily convert the size to a machine
+		// dependent integer type, which causes images larger than 2^31 bytes to
+		// display negative sizes on 32-bit machines:
+		size, err := strconv.ParseInt(string(buf), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		img.Size = int64(size)
+	}
+
+	return img, nil
+}
+
+// saveSize stores the `size` in the provided graph `img` directory `root`.
+func (graph *Graph) saveSize(root string, size int) error {
+	if err := ioutil.WriteFile(filepath.Join(root, "layersize"), []byte(strconv.Itoa(size)), 0600); err != nil {
+		return fmt.Errorf("Error storing image size in %s/layersize: %s", root, err)
+	}
+	return nil
+}
+
+// SetCheckSum sets the checksum for the image layer to the provided value.
+func (graph *Graph) SetCheckSum(id, checksum string) error {
```

Comment:
GetCheckSum should also be made type safe as well. If it is expected GetCheckSum will return a valid Digest, then it will also need to return an error when there is no CheckSum.